### ISSUE TITLE
Option to skip extras requires

### ISSUE THREAD

**Issue by jakirkham** [pypa/setuptools#668]
Currently we find ourselves needing to patch the `setup.py` file of `setuptools` as it has `extras_require` and `dependency_links`, which don't agree with our package manager. We can provide these during the build. However, we find that `setuptools` tries to download them anyways. It would be nice if we had an option to disable these settings during the install or at least recognize that they are already accounted for.

cc @msarahan @pelson @ocefpaf

**Comment by jakirkham** [comment]
See this PR ( https://github.com/pypa/setuptools/pull/721 ) for a potential solution to this problem.

**Comment by jaraco** [comment]
Setuptools doesn't try to download these when I build/install it. Can you share more about what use case is triggering the undesirable behavior?

**Comment by jaraco** [comment]
I'm not sure I follow. I understand you want to disable setuptools from doing any downloads or installs. I'm not sure you can achieve this goal in the general case, because many projects rely on features like `setup_requires` to load requirements at build time.

I still don't understand under what circumstances the extras_requires declared in setuptools are being invoked in the conda environment.

**Comment by jakirkham** [comment]
So, we actually don't find too many issues with our current strategy of disabling downloads (excepting this case with `setuptools`). As we are able to specify [`build` requirements]( http://conda.pydata.org/docs/building/meta-yaml.html#build ) to `conda`, we can satisfy `setup_requires` before any build step has occurred. Again it would be nice if we can agree on some better way to suppress downloads by `setuptools` (perhaps some option provided in a config file). However, this is a separate discussion and probably merits its own issue.

> I still don't understand under what circumstances the extras_requires declared in setuptools are being invoked in the conda environment.

When I started writing this, I was trying to demonstrate and thoroughly explore the problem. However, it is now clear that adding the arguments `--single-version-externally-managed --record=record.txt` makes this unnecessary. I decided to keep this in the post here as it may be of historic value.

Sure, let me provide an example. This PR ( https://github.com/conda-forge/setuptools-feedstock/pull/5 ) drops our [`setup.patch`]( https://github.com/conda-forge/setuptools-feedstock/blob/cf72c792245cc197a7e497bb5fa328223f49c2d6/recipe/setup.patch ), which was used to remove `extras_require` and `dependency_links` from `setup.py`. As a result, that PR is not doing this removal. Taking the OS X Python 2.7 [build]( https://travis-ci.org/conda-forge/setuptools-feedstock/jobs/150162167#L741 ) as an example, we see that installation is blocked by our [`nodownload.patch`]( https://github.com/conda-forge/setuptools-feedstock/blob/cf72c792245cc197a7e497bb5fa328223f49c2d6/recipe/nodownload.patch ).

Now I can see a couple possible causes. One `nodownload.patch` is disabling a download that is being attempted. Two `nodownload.patch` is blocking some other case that is not causing a download.

After trying to remove `nodownload.patch` in PR ( https://github.com/conda-forge/setuptools-feedstock/pull/6 ), it seems that we are running into the second case. This can be seen as `certifi` and, when relevant `wincertstore`, are not downloaded. While this raises some interesting questions about the code path and what `nodownload.patch` is actually doing here, it doesn't solve our problem.

In the end, it seems that providing the flags ( `--single-version-externally-managed --record=record.txt` ) as we do with other Python packages using `setuptools` will solve this problem. This can be seen in PR ( https://github.com/conda-forge/setuptools-feedstock/pull/7 ). It is a bit interesting that these flags are actually needed when no dependencies are (or should be) pulled in. Perhaps this is simply another way to look at some yet again unexpected behavior from the `nodownload.patch`. In any event, it is perfectly reasonable for us to add these flags (we probably should have done so a long time ago) and it does resolve this issue.

**Comment by jaraco** [comment]
Thanks for the detailed writeup, and I'm glad you found a solution. The `single-version-externally-managed` flag was designed to essentially bypass the setuptools installer functionality, so that might in fact be the feature you were seeking.

**Comment by jakirkham** [comment]
So I'm not surprised that this feature works with `install_requires` or `setup_requires`. Though I'm a little surprised that was needed for `extras_require`. The fact that it is needed in this latter case when it shouldn't be downloading anything (and doesn't seem to be) is a little strange. Still it is totally reasonable for us to have these flags; so, am not terribly worried about it.
